File: common/src/object/chunk.rs
pub struct Chunk {
	blocks: Vec<u64>,
	block_types: Vec<(u32, i32)>,
	bits_per_block: u8,
	block_index_shift: u8,
	block_mask: u32,
	clear_mask: u64,
}

impl Chunk {
	pub fn get_block(&self, x: u32, y: u32, z: u32) -> u32 {
		self.get_block_(x & 31 | (y & 31) << 5 | (z & 31) << 10)
	}

	pub fn set_block(&mut self, x: u32, y: u32, z: u32, block: u32) {
		if self.bits_per_block == 0 && block == self.block_types[0].0 {
			return;
		}
		let pos = x & 31 | (y & 31) << 5 | (z & 31) << 10;
		let current_block = self.get_block_(pos);
		if block == current_block {
			return;
		}
		self.get_block_id_and_modify_count(current_block, -1);
		let id = self.get_block_id_and_modify_count(block, 1);
		let offset = (pos & self.block_mask) * self.bits_per_block as u32;
		self.blocks[(pos >> self.block_index_shift) as usize] &= !(self.clear_mask << offset);
		self.blocks[(pos >> self.block_index_shift) as usize] |= (id as u64) << offset;
	}

	fn get_block_(&self, pos: u32) -> u32 {
		if self.bits_per_block == 0 {
			return self.block_types[0].0
		}
		
		self.block_types[(self.blocks[(pos >> self.block_index_shift) as usize] >> ((pos & self.block_mask) * self.bits_per_block as u32) & self.clear_mask) as usize].0
	}

	fn get_block_id_and_modify_count(&mut self, block: u32, count_modifier: i32) -> u32 {
		for i in 0..self.block_types.len() {
			if self.block_types[i].0 == block {
				self.block_types[i].1 += count_modifier;
				if self.block_types[i].1 == 0 && i != 0 {
					let last = self.block_types.len() - 1;
					self.block_types.swap_remove(i);
					for pos in 0..32768 {
						if self.get_block_(pos) == last as u32 {
							let offset = (pos & self.block_mask) * self.bits_per_block as u32;
							self.blocks[(pos >> self.block_index_shift) as usize] &= !(self.clear_mask << offset);
							self.blocks[(pos >> self.block_index_shift) as usize] |= (i as u64) << offset;
						}
					}
					if self.block_types.len() == 1 {
						self.blocks.clear();
						self.bits_per_block = 0;
					}
					else if self.blocks.len() < (1 << (self.bits_per_block >> 1)) {
						self.repack(self.bits_per_block >> 1);
					}
				}
				return i as u32
			}
		}

		let id = self.block_types.len();
		if id >= 1 << self.bits_per_block {
			self.repack(if self.bits_per_block == 0 {1} else {self.bits_per_block << 1});
		}
		self.block_types.push((block, count_modifier));
		id as u32
	}

	fn repack(&mut self, bits_per_block: u8) {
		let blocks_per_value = 64 / bits_per_block;
		let new_block_index_shift = blocks_per_value.trailing_zeros() as u8;
		let new_block_mask = blocks_per_value as u32 - 1;
		let new_clear_mask = (1 << bits_per_block) - 1;

		let old_blocks = std::mem::take(&mut self.blocks);

		self.blocks = vec!(0; 512 * bits_per_block as usize);

		if self.bits_per_block > 0 {
			for i in 0..32768 {
				let b = old_blocks[(i >> self.block_index_shift) as usize] >> ((i & self.block_mask) * bits_per_block as u32) & self.clear_mask;
				if b == 0 {
					continue;
				}
				let offset = (i & new_block_mask) * bits_per_block as u32;
				self.blocks[(i >> new_block_index_shift) as usize] |= b << offset;
			}
		}
		else {
			for i in 0..32768 {
				let offset = (i & new_block_mask) * bits_per_block as u32;
				self.blocks[(i >> new_block_index_shift) as usize] |= (self.block_types[0].0 as u64) << offset;
			}
		}

		self.bits_per_block = bits_per_block;
		self.block_index_shift = new_block_index_shift;
		self.block_mask = new_block_mask;
		self.clear_mask = new_clear_mask;
	}
}

File: src/common/universe/block/Chunk.cpp
#include "Chunk.h"

#include <mutex>

namespace voxel_game::universe::block {
	Chunk::Chunk(const glm::ivec3 &pos): mPos(pos) {}

	uint32_t Chunk::getBlock(const glm::ivec3 &pos) const {
		std::shared_lock lock(mMutex);
		return getBlock_(getBlockIndex(pos));
	}

	void Chunk::setBlock(const glm::ivec3 &pos, const uint32_t block) {
		std::unique_lock lock(mMutex);
		if (mBitsPerBlock == 0 && block == mBlockTypes[0].first) {
			return;
		}
		const uint32_t index = getBlockIndex(pos);
		const uint32_t oldBlock = getBlock_(index);
		if (oldBlock == block) {
			return;
		}
		getOrCreateBlockIDAndModifyCount(oldBlock, -1);
		const uint32_t id = getOrCreateBlockIDAndModifyCount(block, 1);
		const uint32_t offset = (index & mBlockMask) * mBitsPerBlock;
		mBlocks[index >> mBlockIndexShift] &= ~(mClearMask << offset);
		mBlocks[index >> mBlockIndexShift] |= static_cast<uint64_t>(id) << offset;
	}

	glm::ivec3 Chunk::getPos() const {
		return mPos;
	}

	uint32_t Chunk::getBlockIndex(const glm::ivec3& pos) {
		return pos.x & 31 | (pos.y & 31) << 5 | (pos.z & 31) << 10;
	}

	uint32_t Chunk::getBlock_(const uint32_t index) const {
		if (mBitsPerBlock == 0) {
			return mBlockTypes[0].first;
		}
		const uint32_t blockTypeIndex = mBlocks[index >> mBlockIndexShift] >> ((index & mBlockMask) * mBitsPerBlock) & mClearMask;
		if (blockTypeIndex >= mBlockTypes.size()) {
			return 0;
		}
		return mBlockTypes[blockTypeIndex].first;
	}

	uint32_t Chunk::getOrCreateBlockIDAndModifyCount(uint32_t block, int32_t countModifier) {
		for (uint32_t i = 0; i < mBlockTypes.size(); i++) {
			if (mBlockTypes[i].first == block) {
				mBlockTypes[i].second = static_cast<int16_t>(mBlockTypes[i].second + countModifier);
				if (mBlockTypes[i].second <= 0 && i != 0) {
					const uint32_t last = mBlockTypes.size() - 1;
					std::swap(mBlockTypes[i], mBlockTypes[last]);
					mBlockTypes.pop_back();
					for (uint32_t pos = 0; pos < 32768; pos++) {
						if (getBlock_(pos) == last) {
							const uint32_t offset = (pos & mBlockMask) * mBitsPerBlock;
							mBlocks[pos >> mBlockIndexShift] &= ~(mClearMask << offset);
							mBlocks[pos >> mBlockIndexShift] |= i << offset;
						}
					}
					if (mBlockTypes.size() == 1) {
						mBlocks.clear();
						mBitsPerBlock = 0;
					}
					else if (mBlockTypes.size() < 1ULL << (mBitsPerBlock >> 1)) {
						repack(mBitsPerBlock >> 1);
					}
				}
				return i;
			}
		}

		const uint32_t id = mBlockTypes.size();
		if (id >= 1ULL << mBitsPerBlock) {
			repack(mBitsPerBlock == 0? 1: mBitsPerBlock << 1);
		}
		mBlockTypes.emplace_back(block, countModifier);
		return id;
	}

	void Chunk::repack(const uint8_t bitsPerBlock) {
		const uint8_t blocksPer64Bits = 64 / bitsPerBlock;
		const uint8_t newBlockIndexShift = std::countr_zero(blocksPer64Bits);
		const uint32_t newBlockMask = blocksPer64Bits - 1;
		const uint64_t newClearMask = (1ULL << bitsPerBlock) - 1;

		const std::vector<uint64_t> oldBlocks = std::move(mBlocks);

		mBlocks.assign(512 * bitsPerBlock, 0);

		if (mBitsPerBlock > 0) {
			for (uint32_t i = 0; i < 32768; i++) {
				const uint32_t b = oldBlocks[i >> mBlockIndexShift] >> ((i & mBlockMask) * mBitsPerBlock) & mClearMask;
				if (b == 0) {
					continue;
				}
				const uint32_t offset = (i & newBlockMask) * bitsPerBlock;
				mBlocks[i >> newBlockIndexShift] |= static_cast<uint64_t>(b) << offset;
			}
		}
		else {
			for (uint32_t i = 0; i < 32768; i++) {
				const uint32_t offset = (i & newBlockMask) * bitsPerBlock;
				mBlocks[i >> newBlockIndexShift] |= static_cast<uint64_t>(mBlockTypes[0].first) << offset;
			}
		}

		mBitsPerBlock = bitsPerBlock;
		mBlockIndexShift = newBlockIndexShift;
		mBlockMask = newBlockMask;
		mClearMask = newClearMask;
	}
}
